crates/ra_parser/src/grammar.rs

```diff
@@ -74,6 +74,14 @@ pub(crate) fn reparser(
     Some(res)
 }
 
+pub(crate) fn partial_parser(node: SyntaxKind) -> Option<fn(&mut Parser)> {
```

Comment:
Instead of matching over SyntaxKind and unwrapping, let's just introduce a bunch of functions like

```
parse_path
parse_expr
```

etc.

Unlike with reparsing, the caller is in control of the node type that should be parsed .